File: ios/Runner/SocialSharePlugin.swift
import Flutter
import UIKit

/// Native side of `com.timingtap.timing_tap/social_share` (architecture v5
/// §5/§10) — the iOS mirror of `SocialSharePlugin.kt`. Same wire contract
/// (channel name, method names, argument keys, error codes) as the Android
/// side, so the Dart-side `SocialShareService`/`ShareTargetSheet` need zero
/// changes to work on either platform.
///
/// iOS has no `Intent`/`resolveActivity` equivalent: the install probe is
/// `canOpenURL` against each app's own custom URL scheme, and the actual
/// share is Meta's documented "Sharing to Stories" mechanism — write the
/// sticker image + background colors to `UIPasteboard`, then deep-link into
/// the app via that same scheme. No listener/callback registered here
/// either, matching the Android side's one-shot-only shape (architecture
/// §11).
final class SocialSharePlugin: NSObject {

    static let channelName = "com.timingtap.timing_tap/social_share"

    // Wire-format target names — must match `ShareTarget.name` verbatim on
    // the Dart side (`lib/features/sharing/domain/share_target.dart`) and
    // the `TARGET_*` constants in Android's `SocialSharePlugin.kt`.
    private static let targetInstagram = "instagramStory"
    private static let targetWhatsApp = "whatsappStatus"
    private static let targetFacebook = "facebookStory"

    // Custom URL schemes, used both as the install probe (`canOpenURL`) and
    // as the actual deep link fired after the pasteboard write.
    private static let schemeInstagram = "instagram-stories://share"
    private static let schemeFacebook = "facebook-stories://share"

    // WhatsApp's shape is deliberately different from Instagram/Facebook's
    // above (ground truth: `fbsamples/whatsapp_status_api_ios`'s
    // `ContentViewModel.swift`): the install probe is the bare `whatsapp://`
    // scheme, NOT a `-stories://share`-style scheme, and the actual launch
    // afterwards is a `https://` universal link, not a custom-scheme deep
    // link — [whatsappProbeScheme] and [whatsappLaunchURLString] are
    // deliberately different strings; don't conflate them.
    private static let whatsappProbeScheme = "whatsapp://"
    private static let whatsappLaunchURLString = "https://wa.me/media-share-external"
    private static let whatsappPasteboardKey = "com.whatsapp.media-share-external.data"

    func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "installedTargets":
            result(installedTargets())
        case "shareToStory":
            shareToStory(call: call, result: result)
        default:
            result(FlutterMethodNotImplemented)
        }
    }

    /// Which of the 3 targets currently resolve on-device.
    ///
    /// WhatsApp IS included here — reversing an earlier, incorrect
    /// assumption that it had no iOS share path. It has a real one (see
    /// [shareToWhatsAppStatus]); its install probe is just a different shape
    /// from Instagram/Facebook's: the bare `whatsapp://` scheme rather than
    /// a `<brand>-stories://share`-style scheme.
    private func installedTargets() -> [String] {
        var targets: [String] = []
        if let url = URL(string: Self.schemeInstagram), UIApplication.shared.canOpenURL(url) {
            targets.append(Self.targetInstagram)
        }
        if let url = URL(string: Self.whatsappProbeScheme), UIApplication.shared.canOpenURL(url) {
            targets.append(Self.targetWhatsApp)
        }
        if let url = URL(string: Self.schemeFacebook), UIApplication.shared.canOpenURL(url) {
            targets.append(Self.targetFacebook)
        }
        return targets
    }

    private func shareToStory(call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard
            let args = call.arguments as? [String: Any],
            let target = args["target"] as? String,
            let stickerPath = args["stickerPath"] as? String,
            let topColor = args["topColor"] as? String,
            let bottomColor = args["bottomColor"] as? String
        else {
            result(FlutterError(code: "BAD_ARGS", message: "Missing required argument", details: nil))
            return
        }
        let fbAppId = (args["fbAppId"] as? String) ?? ""

        switch target {
        case Self.targetWhatsApp:
            shareToWhatsAppStatus(
                stickerPath: stickerPath,
                topColor: topColor,
                bottomColor: bottomColor,
                result: result
            )
        case Self.targetInstagram:
            shareViaPasteboard(
                scheme: Self.schemeInstagram,
                fbAppId: fbAppId,
                stickerPath: stickerPath,
                topColor: topColor,
                bottomColor: bottomColor,
                stickerImageKey: "com.instagram.sharedSticker.stickerImage",
                topColorKey: "com.instagram.sharedSticker.backgroundTopColor",
                bottomColorKey: "com.instagram.sharedSticker.backgroundBottomColor",
                appIdPasteboardKey: nil,
                result: result
            )
        case Self.targetFacebook:
            // The `com.facebook.sharedSticker.*` pasteboard key spelling
            // below is confirmed against Meta's official "Sharing to Stories
            // — iOS developers" documentation
            // (developers.facebook.com/docs/sharing/sharing-to-stories/ios-developers/),
            // not just inferred from the Android extras' naming.
            //
            // Unlike Instagram, Facebook's contract takes the app ID via a
            // pasteboard key (`com.facebook.sharedSticker.appID`), not a
            // `source_application` URL query param — the deep link itself is
            // the bare `facebook-stories://share` with no query string
            // (code-reviewer flag: the two were wrongly given the same
            // shape in an earlier pass, which would have silently failed).
            shareViaPasteboard(
                scheme: Self.schemeFacebook,
                fbAppId: fbAppId,
                stickerPath: stickerPath,
                topColor: topColor,
                bottomColor: bottomColor,
                stickerImageKey: "com.facebook.sharedSticker.stickerImage",
                topColorKey: "com.facebook.sharedSticker.backgroundTopColor",
                bottomColorKey: "com.facebook.sharedSticker.backgroundBottomColor",
                appIdPasteboardKey: "com.facebook.sharedSticker.appID",
                result: result
            )
        default:
            result(FlutterError(code: "UNKNOWN_TARGET", message: "Unrecognized target: \(target)", details: nil))
        }
    }

    /// Shared by Instagram/Facebook (architecture §3: the outcome card PNG
    /// is always the STICKER layer over a tier-colored gradient background,
    /// never the background asset itself, for every target on both
    /// platforms) — the pasteboard-based equivalent of the Android side's
    /// `TARGET_INSTAGRAM`/`TARGET_FACEBOOK` Intent-extra branches.
    ///
    /// [appIdPasteboardKey] distinguishes the two targets' genuinely
    /// different app-ID contracts (code-reviewer flag — these are NOT the
    /// same shape): Instagram takes the app ID as a `source_application`
    /// query param on the deep-link URL itself (`appIdPasteboardKey ==
    /// nil`); Facebook instead expects it written into the pasteboard item
    /// under `com.facebook.sharedSticker.appID`, with a bare
    /// `facebook-stories://share` URL and no query string.
    private func shareViaPasteboard(
        scheme: String,
        fbAppId: String,
        stickerPath: String,
        topColor: String,
        bottomColor: String,
        stickerImageKey: String,
        topColorKey: String,
        bottomColorKey: String,
        appIdPasteboardKey: String?,
        result: @escaping FlutterResult
    ) {
        guard let probeUrl = URL(string: scheme), UIApplication.shared.canOpenURL(probeUrl) else {
            // Defence-in-depth only, mirroring the Android side's own
            // comment: `ShareTargetSheet` already pre-checks install state
            // before a tile can reach here — only reachable via a race (the
            // app was removed between the probe and the tap).
            result(FlutterError(code: "NOT_INSTALLED", message: "\(scheme) did not resolve", details: nil))
            return
        }

        // Constructed and validated BEFORE anything touches the pasteboard
        // (code-reviewer fix): a malformed link must fail here with nothing
        // written, not after the card image is already sitting in the
        // general pasteboard for no reason. Instagram's app ID rides the
        // URL itself; Facebook's rides the pasteboard item instead (see the
        // doc comment above), so the bare scheme needs no query string.
        let deepLinkString = appIdPasteboardKey == nil
            ? "\(scheme)?source_application=\(fbAppId)"
            : scheme
        guard let deepLink = URL(string: deepLinkString) else {
            result(FlutterError(code: "BAD_ARGS", message: "Malformed deep link", details: nil))
            return
        }

        // Bounded, short-lived allocation on the main thread: card PNGs are
        // small and Share is a rare, user-initiated action, so this doesn't
        // conflict with the app's RAM-resident design principle enough to
        // warrant dispatching the read off-thread.
        guard let imageData = FileManager.default.contents(atPath: stickerPath) else {
            result(FlutterError(code: "BAD_FILE", message: "Could not read sticker file", details: nil))
            return
        }

        var pasteboardItem: [String: Any] = [
            stickerImageKey: imageData,
            topColorKey: topColor,
            bottomColorKey: bottomColor,
        ]
        if let appIdPasteboardKey {
            pasteboardItem[appIdPasteboardKey] = fbAppId
        }
        // Meta's documented convention for both Instagram and Facebook
        // Stories sharing: a short-lived pasteboard write (5 minutes), not a
        // persistent one — this is share-once, ephemeral data, not
        // something that should linger in the general pasteboard
        // indefinitely. `.localOnly` keeps it off Universal Clipboard/
        // Handoff too, so the card image never syncs to the player's other
        // Apple devices (code-reviewer privacy flag).
        let expiration = Date().addingTimeInterval(5 * 60)
        UIPasteboard.general.setItems(
            [pasteboardItem],
            options: [
                .expirationDate: expiration,
                .localOnly: true,
            ]
        )

        UIApplication.shared.open(deepLink, options: [:]) { success in
            if !success {
                // The share never actually launched (e.g. a race where the
                // app was removed between the probe above and this call) —
                // don't leave the card image sitting in the general
                // pasteboard for the full 5-minute expiration when nothing
                // is going to consume it (code-reviewer privacy flag).
                UIPasteboard.general.items = []
            }
            // Mirrors the Android side's `result.success(true)`/
            // `ACTIVITY_NOT_FOUND` shape closely enough for the Dart-side
            // contract: `SocialShareService` already treats any non-`true`
            // return the same as an activity-not-found failure, so a plain
            // `false` here (rather than a synthesized error code) is enough
            // — there's no iOS equivalent of a thrown
            // `ActivityNotFoundException` to catch.
            result(success)
        }
    }

    /// WhatsApp's iOS "Sharing to Status" mechanism (ground truth:
    /// `fbsamples/whatsapp_status_api_ios`'s `ContentViewModel.swift`) — a
    /// deliberately different pasteboard/launch shape from
    /// [shareViaPasteboard]'s Instagram/Facebook contract above: ONE
    /// pasteboard item under [whatsappPasteboardKey], whose value is a
    /// single `[String: Any]` dictionary (not one pasteboard entry per key),
    /// and a `https://` universal-link launch
    /// ([whatsappLaunchURLString]) rather than a custom-scheme deep link —
    /// note this is NOT the same URL as [whatsappProbeScheme], used only for
    /// the `canOpenURL` install probe.
    private func shareToWhatsAppStatus(
        stickerPath: String,
        topColor: String,
        bottomColor: String,
        result: @escaping FlutterResult
    ) {
        guard
            let probeUrl = URL(string: Self.whatsappProbeScheme),
            UIApplication.shared.canOpenURL(probeUrl)
        else {
            // Defence-in-depth only, mirroring shareViaPasteboard()'s own
            // comment — only reachable via a race (the app was removed
            // between the probe and the tap).
            result(FlutterError(code: "NOT_INSTALLED", message: "\(Self.whatsappProbeScheme) did not resolve", details: nil))
            return
        }

        guard let launchUrl = URL(string: Self.whatsappLaunchURLString) else {
            result(FlutterError(code: "BAD_ARGS", message: "Malformed deep link", details: nil))
            return
        }

        guard let imageData = FileManager.default.contents(atPath: stickerPath) else {
            result(FlutterError(code: "BAD_FILE", message: "Could not read sticker file", details: nil))
            return
        }

        guard
            let topComponents = rgbaComponents(fromHex: topColor),
            let bottomComponents = rgbaComponents(fromHex: bottomColor)
        else {
            result(FlutterError(code: "BAD_ARGS", message: "Malformed color", details: nil))
            return
        }

        let payload: [String: Any] = [
            "bundle_id": Bundle.main.bundleIdentifier ?? "",
            "foreground_media": imageData,
            "foreground_media_type": "png",
            // WhatsApp's contract wants one flat RGBA fallback color
            // alongside the 2-stop gradient below; the top gradient stop is
            // the natural choice since it's this app's "primary" tier color.
            "background_color": topComponents,
            "background_gradient": [topComponents, bottomComponents],
        ]

        // WhatsApp's own sample uses a 2-minute pasteboard expiration —
        // deliberately different from Instagram/Facebook's 5 minutes above,
        // not unified. `.localOnly` for the same privacy reasoning as
        // [shareViaPasteboard]: never syncs to Universal Clipboard/Handoff.
        let expiration = Date().addingTimeInterval(120)
        UIPasteboard.general.setItems(
            [[Self.whatsappPasteboardKey: payload]],
            options: [
                .expirationDate: expiration,
                .localOnly: true,
            ]
        )

        // `.universalLinksOnly: true` (code-reviewer flag): without it, a
        // device where WhatsApp hasn't (or can no longer) claim the
        // `wa.me` universal link falls back to opening the URL in Safari
        // instead — and the completion handler still reports `success ==
        // true`, since *opening a URL* succeeded even though the intended
        // app never launched. That would make this method report success
        // back to the Dart layer (sheet dismisses silently) while the
        // player actually lands in a browser with the card sitting unread
        // in the pasteboard. Failing closed here converts that into the
        // same `NOT_INSTALLED`-equivalent `false` the [shareViaPasteboard]
        // Instagram/Facebook path already reports for its own "app didn't
        // actually open" case.
        UIApplication.shared.open(launchUrl, options: [.universalLinksOnly: true]) { success in
            if !success {
                // The share never actually launched — don't leave the card
                // image sitting in the general pasteboard for the full
                // 2-minute expiration when nothing is going to consume it
                // (same privacy reasoning as shareViaPasteboard()).
                UIPasteboard.general.items = []
            }
            result(success)
        }
    }

    /// Parses a `#RRGGBB` hex string (this channel's one canonical
    /// wire-format color — see `share_composition.dart`'s `shareColorHex`)
    /// into an RGBA `[CGFloat]` component array, alpha hardcoded to `1.0` —
    /// the shape WhatsApp's iOS Status pasteboard contract expects for both
    /// `background_color` and each entry of `background_gradient`. Internal
    /// (not `private`), not `fileprivate`/exposed further than needed: kept
    /// directly unit-testable from `RunnerTests` via `@testable import
    /// Runner` — this is the one helper in this file where a silent
    /// off-by-one/format bug would corrupt a shared card's background color
    /// on a real device with no crash to catch it (code-reviewer flag).
    func rgbaComponents(fromHex hex: String) -> [CGFloat]? {
        var sanitized = hex
        if sanitized.hasPrefix("#") {
            sanitized.removeFirst()
        }
        guard sanitized.count == 6, let value = UInt32(sanitized, radix: 16) else {
            return nil
        }
        let r = CGFloat((value & 0xFF0000) >> 16) / 255.0
        let g = CGFloat((value & 0x00FF00) >> 8) / 255.0
        let b = CGFloat(value & 0x0000FF) / 255.0
        return [r, g, b, 1.0]
    }
}

File: android/app/src/main/kotlin/com/timingtap/timing_tap/SocialSharePlugin.kt
package com.timingtap.timing_tap

import android.content.ActivityNotFoundException
import android.content.Context
import android.content.Intent
import android.net.Uri
import androidx.core.content.FileProvider
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import java.io.File

/**
 * Native side of `com.timingtap.timing_tap/social_share` (architecture v5
 * §5/§10): a `MethodChannel` handler covering exactly two responsibilities —
 * probing which of the 3 restricted share targets (Instagram Story, WhatsApp
 * Status, Facebook Story) currently resolve on-device, and firing a single
 * target's direct Story/Status intent.
 *
 * Deliberately tiny (architecture §5 point 5: "roughly 120 lines of
 * Kotlin"), and deliberately retains nothing beyond a single method call:
 * no listener, no `BroadcastReceiver`, no callback registration — one-shot
 * `startActivity` only (architecture §11).
 *
 * [context] is `MainActivity` itself (see `MainActivity.configureFlutterEngine`),
 * not `applicationContext` — a latent Activity-retention risk in general,
 * but bounded today: `FlutterActivity` owns this plugin instance's entire
 * lifetime 1:1 (a fresh `SocialSharePlugin` per `configureFlutterEngine`
 * call, never cached/reused across engine attaches), so this instance never
 * outlives the Activity it holds. Revisit if this plugin is ever registered
 * against a long-lived cached `FlutterEngine` instead.
 */
class SocialSharePlugin(private val context: Context) : MethodChannel.MethodCallHandler {

    companion object {
        const val CHANNEL = "com.timingtap.timing_tap/social_share"

        private const val PKG_INSTAGRAM = "com.instagram.android"
        private const val PKG_WHATSAPP = "com.whatsapp"
        private const val PKG_FACEBOOK = "com.facebook.katana"

        // Wire-format target names — must match `ShareTarget.name` verbatim
        // on the Dart side (`lib/features/sharing/domain/share_target.dart`).
        private const val TARGET_INSTAGRAM = "instagramStory"
        private const val TARGET_WHATSAPP = "whatsappStatus"
        private const val TARGET_FACEBOOK = "facebookStory"

        private val ALL_TARGETS = listOf(TARGET_INSTAGRAM, TARGET_WHATSAPP, TARGET_FACEBOOK)
    }

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        when (call.method) {
            "installedTargets" -> result.success(installedTargets())
            "shareToStory" -> shareToStory(call, result)
            else -> result.notImplemented()
        }
    }

    /**
     * The bare action/type/package shape of each target's intent, with no
     * data/extras yet — shared between the install probe (which only needs
     * this much to call `resolveActivity`) and [shareToStory] (which adds
     * the sticker/color extras on top of the same base before firing it).
     */
    private fun baseIntent(target: String): Intent? = when (target) {
        TARGET_INSTAGRAM -> Intent("com.instagram.share.ADD_TO_STORY").apply {
            type = "image/*"
            setPackage(PKG_INSTAGRAM)
        }
        TARGET_FACEBOOK -> Intent("com.facebook.stories.ADD_TO_STORY").apply {
            type = "image/*"
            setPackage(PKG_FACEBOOK)
        }
        TARGET_WHATSAPP -> Intent(Intent.ACTION_VIEW, Uri.parse("https://wa.me/status")).apply {
            setPackage(PKG_WHATSAPP)
        }
        else -> null
    }

    private fun packageFor(target: String): String? = when (target) {
        TARGET_INSTAGRAM -> PKG_INSTAGRAM
        TARGET_WHATSAPP -> PKG_WHATSAPP
        TARGET_FACEBOOK -> PKG_FACEBOOK
        else -> null
    }

    /**
     * Plain package-installed check — deliberately NOT `resolveActivity`
     * (see the WhatsApp branch of [installedTargets] for why): a simple
     * `getPackageInfo` lookup, unaffected by Android 12+ per-app web-link
     * domain approval.
     */
    private fun isPackageInstalled(pkg: String): Boolean =
        runCatching { context.packageManager.getPackageInfo(pkg, 0) }.isSuccess

    /**
     * `resolveActivity` returns null both when the target app is genuinely
     * absent from the device AND when its package is missing from this
     * manifest's `<queries>` block (architecture §12 code-reviewer flag 8)
     * — a `<queries>` regression is indistinguishable from "not installed"
     * at this layer. Nothing to disambiguate at runtime; kept in sync by
     * hand against `AndroidManifest.xml`'s 3 `<package>` entries.
     */
    private fun installedTargets(): List<String> {
        return ALL_TARGETS.filter { target ->
            if (target == TARGET_WHATSAPP) {
                // NOT `resolveActivity` against `baseIntent`'s
                // `https://wa.me/status` web intent: since Android 12, web
                // intents are subject to per-app domain approval and can
                // resolve to nothing even when WhatsApp is installed (the
                // user can disable "Open supported links" for WhatsApp in
                // system settings) — Meta's own official sample
                // (`fbsamples/whatsapp_status_api_android`) never calls
                // `resolveActivity` for this at all. A plain
                // package-installed check is the only reliable probe here.
                return@filter isPackageInstalled(PKG_WHATSAPP)
            }
            val intent = baseIntent(target) ?: return@filter false
            // Plain `0`, NOT `MATCH_DEFAULT_ONLY` (architecture §2.1, Meta's
            // own sample code): `MATCH_DEFAULT_ONLY` additionally requires
            // the resolved activity to declare `CATEGORY_DEFAULT`, which
            // these vendor-custom actions
            // (`com.instagram.share.ADD_TO_STORY` etc.) are not guaranteed
            // to declare. Using that flag risks a permanently-dimmed tile
            // even when the app IS installed.
            context.packageManager.resolveActivity(intent, 0) != null
        }
    }

    private fun shareToStory(call: MethodCall, result: MethodChannel.Result) {
        val target = call.argument<String>("target")
        val stickerPath = call.argument<String>("stickerPath")
        val topColor = call.argument<String>("topColor")
        val bottomColor = call.argument<String>("bottomColor")
        val fbAppId = call.argument<String>("fbAppId") ?: ""

        if (target == null || stickerPath == null || topColor == null || bottomColor == null) {
            result.error("BAD_ARGS", "Missing required argument", null)
            return
        }

        val packageName = packageFor(target)
        val intent = baseIntent(target)
        if (packageName == null || intent == null) {
            result.error("UNKNOWN_TARGET", "Unrecognized target: $target", null)
            return
        }

        // Same split as installedTargets(): WhatsApp's presence is checked
        // via plain package-installed (never `resolveActivity` against the
        // `https://wa.me/status` web intent — see that function's comment),
        // Instagram/Facebook keep the plain `0` `resolveActivity` check.
        val stillInstalled = if (target == TARGET_WHATSAPP) {
            isPackageInstalled(packageName)
        } else {
            context.packageManager.resolveActivity(intent, 0) != null
        }
        if (!stillInstalled) {
            // Defence-in-depth only — `ShareTargetSheet` already pre-checks
            // install state before letting a tile reach here at all. Only
            // reachable via a race (app uninstalled between the probe and
            // the tap), so mapped by the Dart side to the same
            // "Couldn't open X" copy as a thrown ActivityNotFoundException,
            // not the dimmed-tile "isn't installed" copy.
            result.error("NOT_INSTALLED", "$packageName did not resolve", null)
            return
        }

        val uri: Uri
        try {
            uri = FileProvider.getUriForFile(
                context,
                "${context.packageName}.shareprovider",
                File(stickerPath),
            )
        } catch (e: Exception) {
            result.error("BAD_FILE", "Could not resolve sticker file URI", null)
            return
        }

        // Composition: the outcome card PNG is passed as the STICKER layer
        // over a tier-colored background gradient, never as the background
        // asset itself (architecture §3) — same shape across all 3 targets.
        // Sticker-only sharing (no background asset) uses `type = "image/*"`
        // rather than `setDataAndType(uri, ...)` for Instagram/Facebook —
        // confirmed correct against `react-native-share`'s
        // `InstagramStoriesShare.java`/`FacebookStoriesShare.java`, which use
        // exactly this in production.
        when (target) {
            TARGET_INSTAGRAM -> {
                intent.putExtra("interactive_asset_uri", uri)
                intent.putExtra("top_background_color", topColor)
                intent.putExtra("bottom_background_color", bottomColor)
                intent.putExtra("source_application", fbAppId)
            }
            TARGET_FACEBOOK -> {
                intent.putExtra("interactive_asset_uri", uri)
                intent.putExtra("top_background_color", topColor)
                intent.putExtra("bottom_background_color", bottomColor)
                intent.putExtra("com.facebook.platform.extra.APPLICATION_ID", fbAppId)
            }
            TARGET_WHATSAPP -> {
                // Sticker-only, no background asset — matches Instagram/
                // Facebook's shape (architecture §3: the card PNG is ALWAYS
                // the sticker/foreground layer, never the background asset,
                // for all 3 targets). `EXTRA_STREAM` (background media) is
                // deliberately NOT set here — that would be the "both
                // background AND sticker" double-asset bug this branch
                // avoids.
                //
                // WhatsApp DOES support a gradient-fill background, matching
                // Instagram/Facebook's treatment above — sourced from
                // `fbsamples/whatsapp_status_api_android`'s `MainActivity.kt`.
                // Its extras just use a different color format: `#AARRGGBB`
                // (8 hex digits, alpha first) rather than Instagram/
                // Facebook's 6-digit `#RRGGBB`, hence [toArgbHex] below —
                // done here, not on the Dart side, so `#RRGGBB` stays the one
                // canonical wire format across the whole channel.
                intent.putExtra("share_type", "SHARE_TO_STATUS")
                intent.putExtra("source_app_package_name", context.packageName)
                intent.putExtra("source_app_name", "Stay Alive")
                intent.putExtra("foreground_media", uri)
                intent.putExtra("color_gradient_top", toArgbHex(topColor))
                intent.putExtra("color_gradient_bottom", toArgbHex(bottomColor))
            }
        }

        // Both mechanisms, deliberately (architecture §12 code-reviewer
        // flag 5) — the intent flag alone is not honoured by every
        // receiving composer, so an explicit grantUriPermission is issued
        // as well, before startActivity.
        //
        // Unlike the intent-flag grant (scoped to the receiving task's
        // lifetime), this explicit grant persists until revoked or a system
        // restart — and `CardRenderer` always writes to the same fixed
        // path, so a stale grant would keep giving a previously-used app
        // read access to every LATER card written to that path too. At most
        // one such grant is ever kept outstanding: revoke the previous one
        // (if any) right before issuing a new one, rather than revoking
        // immediately after this call (which would race a composer that
        // reads the URI asynchronously after `startActivity` returns).
        pendingGrant?.let { (pkg, grantedUri) ->
            context.revokeUriPermission(pkg, grantedUri, Intent.FLAG_GRANT_READ_URI_PERMISSION)
        }
        intent.addFlags(Intent.FLAG_GRANT_READ_URI_PERMISSION)
        context.grantUriPermission(packageName, uri, Intent.FLAG_GRANT_READ_URI_PERMISSION)
        pendingGrant = packageName to uri

        try {
            context.startActivity(intent)
            result.success(true)
        } catch (e: ActivityNotFoundException) {
            result.error("ACTIVITY_NOT_FOUND", e.message, null)
        }
    }

    /**
     * Widens a `#RRGGBB` string (the one canonical wire-format color this
     * channel uses — see `share_composition.dart`'s `shareColorHex`) to
     * WhatsApp's `#AARRGGBB` gradient-extra format by prefixing a
     * fully-opaque alpha. Instagram/Facebook's `top_background_color`/
     * `bottom_background_color` extras stay plain 6-digit `#RRGGBB`; only
     * WhatsApp's `color_gradient_top`/`color_gradient_bottom` extras need
     * this 8-digit alpha-first form (`fbsamples/whatsapp_status_api_android`).
     * Widening happens here, not on the Dart side, so the wire format itself
     * stays platform-neutral.
     */
    private fun toArgbHex(rrggbb: String): String = "#FF${rrggbb.removePrefix("#")}"

    /**
     * The most recent (packageName, uri) pair granted via [shareToStory],
     * revoked right before the next grant is issued (see the comment at the
     * call site) — at most one outstanding grant across this plugin's
     * lifetime, one per `MainActivity`/engine lifetime (architecture §11
     * "one-shot, no lingering access" precedent).
     */
    private var pendingGrant: Pair<String, Uri>? = null
}
